### ISSUE TITLE
Support for Rust

### ISSUE THREAD

**Issue by fabianmeyer** [Glavin001/atom-beautify#358]
Please add support for the Rust language. Rust 1.0 was released on 05-15-2015 and the syntax is stable now.

**Comment by nicokoch** [comment]
[https://github.com/nrc/rustfmt](https://github.com/nrc/rustfmt) might be helpful

**Comment by Glavin001** [comment]
Excellent, @nicokoch! That looks great.

For those interested in making this happen, you can submit a Pull Request. I am still working on complete documentation however I have documented most of the steps required to add a new Language and a new Beautifier: https://github.com/Glavin001/atom-beautify/issues/300#issuecomment-107590998

**Comment by Glavin001** [comment]
I have started development on the Rustfmt beautifier support however I have not tested it and it likely is not in a working state, yet. Feel free to fork Atom Beautify, and try it out for yourself and submit a Pull Request when it's completely working. Thanks.

**Comment by Glavin001** [comment]
**Update**: I will be maintaining this documentation on adding languages and beautifiers: https://github.com/Glavin001/atom-beautify/blob/master/docs/add-languages-and-beautifiers.md

**Comment by mattmaslin** [comment]
Rustfmt works correctly.  The only extra step I had to do was copy the <a href="https://github.com/nrc/rustfmt/blob/master/src/default.toml">default.toml</a> file from the Rustfmt project into my home directory.

**Comment by LaylConway** [comment]
It seems there's a problem with the way rustfmt is used by Beautify. When trying to do a format, I get the following error.
```
C:\Users\Layl\AppData\Local\Temp\tmp1151030-5452-fkq6dy:5:5: 5:13 error: file not found for module `frontend`
C:\Users\Layl\AppData\Local\Temp\tmp1151030-5452-fkq6dy:5 mod frontend;
                                                              ^~~~~~~~
C:\Users\Layl\AppData\Local\Temp\tmp1151030-5452-fkq6dy:5:5: 5:13 help: name the file either frontend.rs or frontend/mod.rs inside the directory "C:\\Users\\Layl\\AppData\\Local\\Temp"
thread '<main>' panicked at 'Box<Any>', C:\Users\Layl\.cargo\git\checkouts\syntex-b2c5d63b3d829888\master\syntex_syntax\src\parse/mod.rs:76
```

It seems that rustfmt is expecting the entire cargo project to be in the location it wants to format it.

**Comment by Drakulix** [comment]
There is another problem. rustfmt supports having a config file next to the source file or further up the directory tree. Because atom-beautify copies the source file to /tmp/ (at least on my system - Arch Linux), rustfmt does not find the config file and just applies the default config. The only "workaround" is moving my config to system root, which applies it globally. Not exactly optimal.

**Comment by Glavin001** [comment]
Please feel free to submit a Pull request for rustfmt support in Atom Beautify.
Applicable beautifier code here: https://github.com/Glavin001/atom-beautify/blob/master/src/beautifiers/rustfmt.coffee
I do not believe the beautifier was tested and is likely incomplete.

**Comment by lhecker** [comment]
https://github.com/Glavin001/atom-beautify/blob/master/src/beautifiers/clang-format.coffee could probably be seen as a reference model how rustfmt needs to be invoked like. I tried to modify the code myself, but gave up after a while since I failed to understand how this project works. Maybe someone with more experience could copy over and integrate the code. :blush:

**Comment by sidred** [comment]
I have working version of Beautifier support for rustfmt.toml at https://github.com/sidred/atom-beautify/tree/rustfmt-update. Waiting for  https://github.com/rust-lang-nursery/rustfmt/pull/780 to be merged before a submititng PR.

**Comment by xavier83** [comment]
@sidred  looks like it has been merged. any update?

**Comment by sidred** [comment]
Merged via #790. Beautifier should read the rustfmt.toml in an file path.

**Comment by imp** [comment]
rustfmt development recently has been moved to rustfmt-nightly branch, which original rustfmt being frozen. As a side effect of that change, the version numbering has been reset back to 0.1.x. However, atom-beautify check for the version being higher than 0.5. Naturally, it comes to the wrong conclusion with rustfmt-nightly as not being good enough. Fortunately, it is easy to distinguish between the two branches - as nightly clearly says so in its `--version` output:
```
$ /usr/local/bin/rustfmt --version
0.1.9-nightly ( )
```
It would be great to add extra check for running newer branch of rustfmt and clear it for using inside atom-beautify

**Comment by Glavin001** [comment]
Pull Requests welcome! See https://github.com/Glavin001/atom-beautify/blob/master/src/beautifiers/rustfmt.coffee for applicable code.

I am not a Rust user/developer, so it is unlikely I would write the Pull Request. However, I will be happy to review and merge any changes. Thanks in advance!

**Comment by szeck87** [comment]
Published to v0.30.5
